### PR TITLE
add options to proj_conv and all_lrelu in the definition of SPADEMaskerDecoder

### PR DESCRIPTION
**Pull request by tianyu-z**

### CODE REVIEW

File: omnigan/blocks.py [hunk 1]
```diff
@@ -554,7 +556,10 @@ def forward(self, x, seg):
         dx = self.conv_1(self.activation(self.norm_1(dx, seg)))
 
         out = x_s + dx
-        return out
+        if self.last_activation == "lrelu":
+            return self.activation(out)
+        elif self.last_activation is None:
+            return out
```
Comment (vict0rsch) [inline omnigan/blocks.py:562]: need to have an `else` which raises a NotImplementedError

File: omnigan/generator.py [hunk 2]
```diff
@@ -382,24 +386,54 @@ def __init__(self, opts):
...
-                self.input_dim[0] * 2,
-                self.z_nc,
-                3,
-                padding=1,
-                activation="lrelu",
-                pad_type="reflect",
-                norm="spectral_batch",
-            )
+            if self.opts.gen.m.use_proj:
+                num_proj = self.opts.gen.m.proj_dim
```
Comment (vict0rsch) [inline omnigan/generator.py:390]: `proj_dim` is a better name for that variable, more obvious

**Review by vict0rsch — CHANGES_REQUESTED** [review]

**Comment by tianyu-z** [comment]
I fixed them but I am doing some other experiments. I would like to keep this PR as a draft now. Will let you know when my other exps are done.

**Comment by vict0rsch** [comment]
Cool thanks :)

**Comment by tianyu-z** [comment]
What this PR does:
Add options to backprop conditions for the maskSPADEdecoder and also add options for proj_conv and all_lrelu as the activation functions.

Conclusion (bs=5):
1. Training Speed(Without Backprop) = 758.1 step/hr; Training Speeing(With Backprop) = 638.79 step/hr. 
2. Use_proj does help overall. But proj_dim = 64 does not have much difference with proj_dim = 128.
3. Models without Backprop get reasonable result on Mask at epoch = 7 and Models with Backprop get reasonable result on Mask at epoch = 5. But after epoch 15, they perform similar in the result of the Masker. ([Comparision exps](https://www.comet.ml/tianyu-z/omnigan/dc81d1f10c1b44ebb6a0b0548f49164e/8bc4c25f34fb44db8b403d0b2bde77e0/compare?experiment-tab=images&groupBy=false&orderBy=asc&search=val_r_m_%5B0-9%5D%2A_%230&searchType=regex&sortBy=step))

**Comment by tianyu-z** [comment]
@vict0rsch I retrained the models. From my viewpoint, the conclusion above still holds:
[use_proj+backprop](https://www.comet.ml/tianyu-z/omnigan/b659c2e62ff2468191898c44d54ec81c?experiment-tab=images&groupBy=false&orderBy=asc&search=val_r_m_%5B0-9%5D%2A_%233&searchType=regex&sortBy=step
)
[backprop](https://www.comet.ml/tianyu-z/omnigan/d48821d41c124373a24b6f241e5e6ad2?experiment-tab=images&groupBy=false&orderBy=desc&search=val_r_m_%5B0-9%5D%2A_%231&searchType=regex&sortBy=step
)
[use_proj](https://www.comet.ml/tianyu-z/omnigan/2562a3588f494bf3a758f805c252e25d?experiment-tab=images&groupBy=false&orderBy=asc&search=val_r_m_%5B0-9%5D%2A_%233&searchType=regex&sortBy=step
)
One more thing that I noticed:
models with proj layer tend to flood more:
image
in an extreme case, no one survives: 
image

models without proj layer tend to flood less:
image
in an extreme case, people almost are not flooded:
image

**Pull request merged**